cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs

```diff
@@ -5,89 +5,477 @@
+
+            var nhsNumberMatchResidentId = DatabaseContext.ResidentEntities.FirstOrDefault(r => r.NhsNumber == matchingNhsNumber)?.Id;
+
+            // act
+            var duplicateResidentId = _classUnderTest.FindResident(searchParameters);
+            bool isResidentDuplicate = duplicateResidentId != null;
+
+            // assert
+            isResidentDuplicate.Should().BeTrue();
+            duplicateResidentId.Should().Be(nhsNumberMatchResidentId);
+        }
+
+        // If NHS number is empty or null or does not match &
+        // If Uprn is present, then:
+        // Resident is considered a duplicate, when First Name, Last Name & Uprn match.
+
+        [TestCase(null)]
+        [TestCase("")]
+        [TestCase(" ")]
+        [TestCase("nhs number")]
+        public void FindResidentReturnsAMatchWhenNoNhsNumberIsPresentAndWhenThereIsAResidentWithMatchingNonEmptyUprnAndFirstAndLastNamesForAllEmptyCasesOfNhsNumber(string testCaseNhsNumber)
+        {
+            // arrange
+            var matchingEmptyNhsNumber = testCaseNhsNumber;
+            var matchingFirstName = _faker.Random.Hash();
+            var matchingLastName = _faker.Random.Hash();
+            var matchingUprn = _faker.Random.Hash();
+
+            var searchParameters = new FindResident
             {
-                Id = existingResident.Id,
-                Uprn = "uprn",
-                FirstName = "FirstName",
-                LastName = "LastName"
+                NhsNumber = matchingEmptyNhsNumber,
+                Uprn = matchingUprn,
+                FirstName = matchingFirstName,
+                LastName = matchingLastName,
             };
 
-            createdRecord.Should().BeEquivalentTo(expectedRecord);
+            // empty nhs number match - making sure that doesn't get treated as a false positive
+            var existingResidentEmptyNhsNumberMatch = new ResidentEntity
+            {
+                NhsNumber = matchingEmptyNhsNumber,
+                Uprn = _faker.Random.Hash(),
+                FirstName = _faker.Random.Hash(),
+                LastName = _faker.Random.Hash()
+            };
+
+            var existingResidentUprnMatch = new ResidentEntity
+            {
+                NhsNumber = _faker.Random.Hash(),
+                Uprn = matchingUprn,
+                FirstName = matchingFirstName,
+                LastName = matchingLastName
+            };
+
+            // testing the no match path (adding this 1 line, so I wouldn't need to copy the whole test)
+            if (matchingEmptyNhsNumber == "nhs number") existingResidentEmptyNhsNumberMatch.NhsNumber = _faker.Random.Hash();
+
+            DatabaseContext.ResidentEntities.Add(existingResidentEmptyNhsNumberMatch);
+            DatabaseContext.ResidentEntities.Add(existingResidentUprnMatch);
+            DatabaseContext.SaveChanges();
+
+            var uprnMatchResidentId = DatabaseContext.ResidentEntities.FirstOrDefault(r => r.Uprn == matchingUprn)?.Id;
```

Comment:
Same as before, can't we access the Id through ```existingResidentUprnMatch.Id``` ?

Reply:
Unfortunately no, it's because we auto-generate Ids, so we don't know what it will be unless we check that. Due to NUnit tests running in a different order than shown (and maybe even asynchronously), it would be hard to reason about the Id that will be given. Even if we guess it, and the hardcode it - any time we add new tests, there's a possibility that our hardcoded guess could fail, due to the order changing order of how the tests run. _(Our tests setup doesn't clear DB properly, but honestly, even if it did, I feel relying on a DB query that's tested Microsoft isn't as bad as hardcoding a guess.)_

Reply:
Nvm... adding a record to DB has a side effect that updates the object that you're inserting. Good shout!